File: integration-test/java/shared/src/main/java/com/qualityminds/lazyval/integration/shared/Birthdate.java
package com.qualityminds.lazyval.integration.shared;

import org.jspecify.annotations.Nullable;

import java.time.LocalDate;
import java.util.Objects;
import java.util.regex.Pattern;

public final class Birthdate implements Comparable<Birthdate> {

    public static final Birthdate UNKNOWN = new Birthdate(new ParsedDate(0, 0, 0));
    private static final Parser parser = new Parser();

    @Override
    public int compareTo(@Nullable Birthdate other) {
        if(other == null){
            return 1;
        }
        return this.value.compareTo(other.value);
    }

    public sealed interface State {
        record Complete(LocalDate date) implements State {}
        record DayUnknown (int month, int year) implements State {}
        record DayMonthUnknown (int year) implements State {}
        record Unknown() implements State {
            private static final Unknown INSTANCE = new Unknown();
        }
    }

    /**
     * ISO-8601 representation
     */
    private final String value;
    private final transient State state;

    public String value(){ return value; }

    public State getState() { return state; }

    private Birthdate(final ParsedDate parsedDate){
        this.value = parsedDate.asIsoString();
        if(parsedDate.isComplete()){
            this.state = new State.Complete(LocalDate.of(parsedDate.year, parsedDate.month, parsedDate.day));
        }else if(parsedDate.isDayUnknown()){
            this.state = new State.DayUnknown(parsedDate.year, parsedDate.month);
        }else if(parsedDate.isDayMonthUnknown()){
            this.state = new State.DayMonthUnknown(parsedDate.year);
        }else{
            this.state = State.Unknown.INSTANCE;
        }
    }

    public static Birthdate of(String value){
        if(value == null || value.isBlank()){
            return null;
        }
        var parsedDate = parser.parse(value);
        // use shared instance
        if(parsedDate.isUnknown()){
            return UNKNOWN;
        }
        return new Birthdate(parsedDate);
    }


    public static Birthdate of(LocalDate date){
        if(date == null){
            return null;
        }
        return new Birthdate(ParsedDate.from(date));
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        Birthdate birthdate = (Birthdate) o;
        return Objects.equals(value, birthdate.value);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(value);
    }

    private record ParsedDate(int year, int month, int day) {
        ParsedDate {
            checkValid();
        }

        static ParsedDate from(LocalDate date){
            return new ParsedDate(date.getYear(), date.getMonthValue(), date.getDayOfMonth());
        }

        boolean isUnknown() {
            return year == 0 && month == 0 && day == 0;
        }

        boolean isComplete(){
            return year != 0 && month != 0 && day != 0;
        }

        boolean isDayUnknown(){
            return year == 0 && month == 0 && day != 0;
        }

        boolean isDayMonthUnknown(){
            return year != 0 && month == 0 && day == 0;
        }

        /**
         * will cause a DateTimeParseException for invalid Dates
         */
        @SuppressWarnings("ResultOfMethodCallIgnored")
        void checkValid(){
            LocalDate.of(
                    year,
                    month != 0 ? month : 1,
                    day != 0 ? day : 1
            );
        }

        /**
         * Converts the parsingresult to yyyy-mm-dd
         */
        String asIsoString(){
            return "%04d-%02d-%02d".formatted(year, month, day);
        }
    }

    private static class Parser {

        private static final Pattern DATE_PATTERN_ISO_8601 = Pattern.compile("^\\d{4}-\\d{2}-\\d{2}$");

        private ParsedDate parse(String value){
            final ParsedDate parsed;
            if(DATE_PATTERN_ISO_8601.matcher(value).matches()){
                parsed = parseIso(value);
            }else{
                throw new IllegalArgumentException("Date '%s' is not matching yyyy-mm-dd".formatted(value));
            }
            return parsed;
        }

        private static ParsedDate parseIso(final String value){
            return new ParsedDate(
                    Integer.parseInt(value.substring(0,4)),
                    Integer.parseInt(value.substring(5,7)),
                    Integer.parseInt(value.substring(8,10))
            );
        }
    }
}

File: integration-test/kotlin/shared/src/main/kotlin/com/qualityminds/lazyval/integration/shared/Birthdate.kt
package com.qualityminds.lazyval.integration.shared

import java.time.LocalDate
import java.util.regex.Pattern

class Birthdate private constructor(parsedDate: ParsedDate) : Comparable<Birthdate> {

    /**
     * ISO-8601 representation
     */
    val value: String = parsedDate.asIsoString()

    @Transient
    val state: State = when {
        parsedDate.isComplete -> State.Complete(LocalDate.of(parsedDate.year, parsedDate.month, parsedDate.day))
        parsedDate.isDayUnknown -> State.DayUnknown(parsedDate.year, parsedDate.month)
        parsedDate.isDayMonthUnknown -> State.DayMonthUnknown(parsedDate.year)
        else -> State.Unknown
    }

    override fun compareTo(other: Birthdate): Int = value.compareTo(other.value)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        return value == (other as Birthdate).value
    }

    override fun hashCode(): Int = value.hashCode()

    sealed interface State {
        data class Complete(val date: LocalDate) : State
        data class DayUnknown(val month: Int, val year: Int) : State
        data class DayMonthUnknown(val year: Int) : State
        data object Unknown : State
    }

    private data class ParsedDate(val year: Int, val month: Int, val day: Int) {

        init {
            checkValid()
        }

        val isUnknown: Boolean get() = year == 0 && month == 0 && day == 0
        val isComplete: Boolean get() = year != 0 && month != 0 && day != 0
        val isDayUnknown: Boolean get() = year == 0 && month == 0 && day != 0
        val isDayMonthUnknown: Boolean get() = year != 0 && month == 0 && day == 0

        /**
         * will cause a DateTimeParseException for invalid Dates
         */
        private fun checkValid() {
            LocalDate.of(
                year,
                if (month != 0) month else 1,
                if (day != 0) day else 1,
            )
        }

        /**
         * Converts the parsingresult to yyyy-mm-dd
         */
        fun asIsoString(): String = "%04d-%02d-%02d".format(year, month, day)

        companion object {
            fun from(date: LocalDate): ParsedDate =
                ParsedDate(date.year, date.monthValue, date.dayOfMonth)
        }
    }

    private object Parser {

        private val DATE_PATTERN_ISO_8601: Pattern = Pattern.compile("^\\d{4}-\\d{2}-\\d{2}$")

        fun parse(value: String): ParsedDate {
            if (!DATE_PATTERN_ISO_8601.matcher(value).matches()) {
                throw IllegalArgumentException("Date '$value' is not matching yyyy-mm-dd")
            }
            return parseIso(value)
        }

        private fun parseIso(value: String): ParsedDate = ParsedDate(
            value.substring(0, 4).toInt(),
            value.substring(5, 7).toInt(),
            value.substring(8, 10).toInt(),
        )
    }

    companion object {
        @JvmField
        val UNKNOWN: Birthdate = Birthdate(ParsedDate(0, 0, 0))

        @JvmStatic
        fun of(value: String?): Birthdate? {
            if (value.isNullOrBlank()) return null
            val parsedDate = Parser.parse(value)
            // use shared instance
            return if (parsedDate.isUnknown) UNKNOWN else Birthdate(parsedDate)
        }

        @JvmStatic
        fun of(date: LocalDate?): Birthdate? {
            if (date == null) return null
            return Birthdate(ParsedDate.from(date))
        }
    }
}
